parmed/structure.py

```diff
@@ -1574,10 +1574,21 @@ def combining_rule(self):
         return self._combining_rule
 
     @combining_rule.setter
-    def combining_rule(self, thing):
+    def combining_rule(self, thing, recalculate_14=False):
```

Comment:
This is a setter, so there's no way to actually specify `recalculate_14=True` as written.  Also, `Atom` has a `sigma_14` attribute that specifies the 1-4 sigma parameter (which may be different than its "normal" sigma parameter).

Instead, I'd do something like this:

```python
@combining_rule.setter
def combining_rule(self, thing):
    if self._combining_rule == thing:
        # nothing to do
        return
    if self.has_NBFIX():
        raise ValueError("Cannot change the combining rule with modified off-diagonal Lennard-Jones parameters")

    if thing == "lorentz":
        combine = lambda sig1, sig2: (sig1 + sig2) / 2
    elif thing == "geometric":
        combine = lambda sig1, sig2: (sig1 * sig2) ** 0.5
    if any(abs(adj.type.sigma - combine(adj.atom1.sigma_14, adj.atom2.sigma_14)) > TINY):
        raise ValueError("Cannot change the combining rule with modified off-diagonal 1-4 Lennard-Jones parameters")
    for adj in structure.adjusts:
        adj.type.sigma = combine(adj.atom1.sigma_14, adj.atom2.sigma_14)
```

Reply:
Good catch! Thank you for the feedback! I only have one question about this line: 
```python3  
if any(abs(adj.type.sigma - combine(adj.atom1.sigma_14, adj.atom2.sigma_14)) > TINY for adj in self.adjusts):
        raise ValueError("Cannot change the combining rule with modified off-diagonal 1-4 Lennard-Jones parameters")
```
it looks like the `if` statement is checking if the newly calculated sigma would be much different than those currently stored in `adj.type.sigma`, but I am not sure why this will raise the ValueError below.

Reply:
Basically what's happening here is we are throwing a ValueError if there are any off-diagonal modifications of 1-4 nonbonded interactions.  The Lennard-Jones parameter matrix is built by computing pairwise "combined" sigma and epsilon properties.  NBFIX (nonbond "fixes") allow you flexibility in fitting the Lennard-Jones potential by changing "off-diagonal" elements of the parameter matrix that *doesn't* obey the normal combining rules (CHARMM uses this to tune specific pairs of atomic interactions without affecting all of the other interactions).

However, if you change the combining rules and "recompute" the sigmas on the 1-4 adjustments, you would wipe out any specific pair overrides (NBFIXes) you implemented.  It's unclear exactly what should be done if there is NBFIX implemented:

* Do you keep the pair-specific parameters used to override the combining rules?
* Or do you overwrite them by applying the combining rules to all pairs?
* Or do you throw an error and force the developer to resolve the issue?

There are reasonable arguments to support either of the first two options (leave them alone or overwrite), and large numbers of people could (justifiably) be surprised by either of the first two behaviors.  So the only safe option is to do the third, which will ensure nobody gets surprised, even if some people are slightly inconvenienced.

But if you go through the effort of NBFIXing certain interactions, you aren't likely to change the combining rule so this safety valve probably shouldn't hit anybody (because nobody should really try to change the combining rule *after* NBFIXing).  But just in case, it's worth throwing an exception to make sure surprising (but subtle and hard-to-detect) doesn't slip through unprotected.

Reply:
Gotcha. I added that checks to the methods

Reply:
Spending some more thought on this, I am still a bit confusing by the the last check (`if any(abs(adj.type.sigma - combine(adj.atom1.sigma_14, adj.atom2.sigma_14)) > TINY for adj in self.adjusts):`). If we are worried about the NBFix implemented by the user, shouldn't it has already captured by the first condition check (`if self.has_NBFIX():`)? 
With the last check being: "Error out if if any of the existing `adj.type.sigma` differ by the newly calculated `sigma`", it kind of defeat the purpose of this PR in updating the sigma if the user update the overall combining rule of the structure.